trieste/acquisition/rule.py

```diff
@@ -211,31 +212,242 @@ def acquire(
+
+        Let's suppose we have P pending points. To optimize the acquisition function
+        we call it with batches of size P+1, where first P points are fixed to pending points.
+        Optimization therefore happens over the last point only, which is returned.
+
+        :param search_space: The local acquisition search space for *this step*.
+        :param datasets: The known observer query points and observations.
+        :param models: The models of the specified ``datasets``.
+        :return: A function that constructs the next acquisition state and the recommended query
+            points from the previous acquisition state.
+        """
+        if self._acquisition_function is None:
+            self._acquisition_function = self._builder.prepare_acquisition_function(
+                datasets, models
+            )
+        else:
+            self._acquisition_function = self._builder.update_acquisition_function(
+                self._acquisition_function, datasets, models
+            )
+
+        def state_func(
+            state: AsynchronousRuleState | None,
+        ) -> tuple[AsynchronousRuleState | None, TensorType]:
+            tf.debugging.Assert(self._acquisition_function is not None, [])
+
+            if state is None:
+                state = AsynchronousRuleState(None)
+
+            state = state.remove_points(datasets[OBJECTIVE].query_points)
+
+            if state.has_pending_points:
+                pending_points: TensorType = state.pending_points
+
+                def function_with_pending_points(x: TensorType) -> TensorType:
+                    # stuff below is quite tricky, and thus deserves an elaborate comment
+                    # we receive unknown number N of points to evaluate
+                    # note that we don't deal with batches here
+                    # so the shape of `x` is [N, 1, D]
+                    # we want to add P pending points to each point
+                    # so that acquisition actually receives N batches of size [P+1, D] each
+                    # therefore here we prepend each point with all pending points
+                    # resulting a shape [N, P+1, D]
+                    # we do that by repeating pending points N times and concatenating with x
+
+                    # pending points are 2D, we need 3D and repeat along first axis
+                    expanded = tf.expand_dims(pending_points, axis=0)
+                    pending_points_repeated = tf.repeat(expanded, [tf.shape(x)[0]], axis=0)
+                    all_points = tf.concat([pending_points_repeated, x], axis=1)
+                    return cast(AcquisitionFunction, self._acquisition_function)(all_points)
+
+                acquisition_function = function_with_pending_points
+            else:
+                # mypy gets confused about types of these functions
+                acquisition_function = self._acquisition_function  # type: ignore
```

Comment:
why not cast it like above? (or alternatively, type ignore the previous cast)

(btw mypy is correct to be confused because  this is happening inside a closure and there's no guarantee that by the time we execute state_func sef._acquisition_function won't be None again)

Reply:
Cast here doesn't work because mypy works out some confusing type for `acquisition_function` based on assignment two lines above. I ended up casting to AcquisitionFunction on both of these lines